[RDY] atomic multi request for async remote plugins

bfredl opened the pull request:
Based on @ZyX-I:s suggestion https://github.com/neovim/neovim/pull/4411#issuecomment-205972405 this allows a rpc client to send a list of requests for evaluation at once. This allows "atomic" execution from an external async context (of course, any request that can lead to the execution of vimscript can and will have side-effects in the middle of executing the multi request). Even if `set_var` and `del_var` is changed to not return the oldvalue as suggested in #4411 the old behavior can be kept by bundling a "get_var" and "set_var" request

To support simple test-and-set behavior, it is possible to assert the return value of a request and abort the execution. For anything more complicated it is better to inject some vimscript (or lua) and let it do it.

The argument is an array of `[methodname, args, sendreturn{, assertreturn}]` entries.
simple example:

```
vim.api.multi_request([['vim_eval', ['3+3'], False, 6],
                       ['vim_get_current_line', [], True]])
```

It could be possible to do something like first

```
get_lines [buffer, -1]
get_linecount [buffer]  
get_var [buffer, "changedtick"]
```

and then

```
get_var [buffer, "changedtick"] False oldchangedtick
set_lines [buffer, linenr, oldtext+newtext]
add_highlight [buffer, linenr, ...]
```

So that the update to the buffer is only done if the buffer was unchanged. The first return value is the number of successful request to this condition can be detected.

Bikeshed of the api and return format welcome. It is quite-low level, but the intention is mainly for client/hosts implementations that can export nicer APIs like  `list.pop` and `dict.pop` in the case of python, to plugins.

bfredl committed:
api: call multiple methods atomically (useful in async contexts)

remove unused response_id parameter of handle_nvim_... helpers

justinmk commented on src/nvim/api/vim.c:
```diff
@@ -607,6 +607,96 @@ Array vim_get_api_info(uint64_t channel_id)
   return rv;
 }
 
+Array vim_multi_request(uint64_t channel_id, Array calls, Error *err)
```
vim_batch_call ?

justinmk commented on src/nvim/api/vim.c:
```diff
@@ -607,6 +607,96 @@ Array vim_get_api_info(uint64_t channel_id)
+    if (call.size != 3 && call.size != 4) {
+      api_set_error(err,
+                    Validation,
+                    _("All items in calls array must be arrays of size 3 or 4"));
```
Consider a fixed size of 4, require nil arg.

bfredl replied:
There is a problem with that, NIL is an actual value to which the return value could be compared. If we insist of a fixed size we either need a bool flag `[name, args, sendret, assertret, assertval]` or maybe merge with `sendret` (because if you assert the value, you don't want it returned, unless the assertion fails).

justinmk commented:
> Bikeshed of the api and return format welcome

What's the return format? The general shape looks reasonable. Granular error reporting (eg of the failed assertionwill)  help a lot.

equalsraf commented:
Have not checked thoroughly, but this makes me wonder if anybody out there hasn't looked at msgpack-rpc "transactions" by placing multiple rpc messages in a new message type.

bfredl commented:
Updated. I removed the preassert for now, this already useful in various cases without it, (such as updating multiple bufhl in an async context without causing excessive redraws, or emulating the old behavior of del_var after #5336)  and that could be added in a separate PR without breaking existing usages.

bfredl commented:
@justinmk I'd like to nominate this for 0.1.6, as said this would be able to fix real performance issues in bufhl (ref https://github.com/arakashic/chromatica.nvim/issues/18), and, say implement `vim.dictonary.pop()` which is missing in the python-client for some reason, without using deprecated methods. (Though both are _in principle_ possible with vimL helpers, it surely would seem strange if consumers of the new API would need to rely on viml code to be performant)

justinmk commented:
Ok I'll look at this today. Hopefully others can too.

oni-link commented on src/nvim/api/vim.c:
```diff
@@ -663,6 +663,110 @@ Array nvim_get_api_info(uint64_t channel_id)
+
+    MsgpackRpcRequestHandler handler = msgpack_rpc_get_handler_for(name.data,
+                                                                   name.size);
+    Object result = handler.fn(channel_id, UINT64_MAX, args, &nested_error);
```
`UINT64_MAX` -> `NO_RESPONSE`

bfredl replied:
I don't like that really, no API function does it yet, but they could use `NO_RESPONSE` for the optimization "no response exected, don't bother calculating the return value". A separate flag here is probably the best (to use by `eval.c:api_wrapper` also)

justinmk replied:
So `INTERNAL_RESPONSE` is a future advisory flag, it's not used any where yet.

oni-link commented on src/nvim/api/vim.c:
```diff
@@ -663,6 +663,110 @@ Array nvim_get_api_info(uint64_t channel_id)
+    Object result = handler.fn(channel_id, UINT64_MAX, args, &nested_error);
+    if (nested_error.set) {
+      // error handled after loop
+      break;
```
Is on error `result` already freed or `NIL`? Otherwise missing clean up.

bfredl replied:
it should be an invariant, whenever an api function sets an error, the return value should be NIL (or at least non-allocating).

oni-link commented on src/nvim/api/vim.c:
```diff
@@ -663,6 +663,110 @@ Array nvim_get_api_info(uint64_t channel_id)
+/// with at least three elements: the request name, an array of arguments, and
+/// a boolean indicating if the return value should be sent back.
+///
+/// @return an array with three elements. The first is the number of calls that
```
In case of a validation error an empty  array is returned.

bfredl replied:
Do you mean validation of the multi_request array itself? Then an error is sent back, nothing can be returned.

oni-link commented on src/nvim/api/vim.c:
```diff
@@ -663,6 +663,110 @@ Array nvim_get_api_info(uint64_t channel_id)
+/// call resulting in an error. The second is an array of return values, for
+/// calls when the third boolean flag was set to true. The third is NIL if all
+/// calls succeeded. If a call resulted in an error, it is a two-element array
+/// with the error code and message.
```
code -> type ?

oni-link commented on src/nvim/api/vim.c:
```diff
@@ -663,6 +663,110 @@ Array nvim_get_api_info(uint64_t channel_id)
+    }
+
+    if (retval) {
+      ADD(results, result);
```
If a call fails but a result should be returned, nothing is put in `results`. Is this expected?

bfredl replied:
yes, the comlete multi_request is then interrupted, so there is no ambiguity what calls the present elements in `results` come from.

oni-link commented on src/nvim/api/vim.c:
```diff
@@ -663,6 +663,110 @@ Array nvim_get_api_info(uint64_t channel_id)
+                    _("name must be String"));
+      goto end;
+    }
+    String name = call.items[0].data.string;
```
`name` and `args` could be moved after the validation code.

bfredl replied:
They could, but to me matching a field from a discriminated union is one operation, which happens to be two steps in language like C, thus best kept together.

bfredl commented:
Thanks, updated.

justinmk commented on src/nvim/api/private/defs.h:
```diff
@@ -30,10 +30,14 @@ typedef enum {
+#define INTERNAL_RESPONSE (UINT64_MAX-1u)
+
+/// Used as channel_id when the call is local.
 #define INVALID_CHANNEL UINT64_MAX
```
I guess this should be named `INTERNAL_CHANNEL`.

bfredl replied:
Well it is not a channel, maybe `INTERNAL_CALL`

justinmk commented on src/nvim/api/vim.c:
```diff
@@ -663,6 +663,115 @@ Array nvim_get_api_info(uint64_t channel_id)
+///
+/// @return an array with three elements. The first is the number of calls that
+/// where performed without errors. If it is the same as the length of "calls",
+/// all calls were successful, otherwise it is the number of the last attempted
```
I'm confused by this. It says the first item is `number of calls performed without errors`, then says if an error occurred the first item is the `call that resulted in an error`.

justinmk commented:
Can we name it `request_atomic`? Because (1) we should keep global functions starting with a verb (`:h dev-api`) unless it is pseudo-namespaced, (2) "atomic" is keyword-friendly and communicates much more information ("multi" which does not advertise the nice properties of this PR).

As @equalsraf  [mentioned](https://github.com/neovim/neovim/pull/4568#issuecomment-222296538) it would seem ideal to have this at the protocol layer (it's a bit weird that clients will be sending "local RPC requests" via the remote RPC API). But the simplicity of this implementation cannot be denied and it will give clients something very useful. So we can't let perfection block this.

bfredl commented on src/nvim/api/private/defs.h:
```diff
@@ -30,10 +30,14 @@ typedef enum {
 
-/// Used as channel_id when the call is local
+/// Used as response id when the call is local or part of a multi-request.
+#define INTERNAL_RESPONSE (UINT64_MAX-1u)
```
When thinking of it, no API method is actually using the response id number, nor would I expect it to be used (possibly in #4971, but then I guess it would be part of the stream, and that would need to change signatures of methods using  anyway), so we could just make this a `bool` for now, or remove it from this PR, it could be introduced when actually used.

bfredl replied:
Actually is is not possible for a API method to receive the `response_id` (like it is with the `channel_id`), it is just an always unused argument to `handle_nvim_[method]`. I'll just remove it for now.

justinmk commented:
Instead of request_atomic, send_atomic_request or send_multi_request?

bfredl commented:
Not sure I like "send" it would sound like nvim should send the requests somewhere else. Though I'm not super happy with "request" either, this could just as well be used as a notification (and reasonably then no return value would be requested). Maybe `call_atomic` or `call_multiple` ?

justinmk commented:
`call_atomic` seems best. Though it's inconsistent with the existing `call` (which only calls VimL functions, not request tuples), there's really no perfect approach here because of the unusual nature of this (we don't have any other API functions that take a request tuple).

bfredl commented:
Updated. An alternative could be `call_methods_atomic` if we use the distinction of API methods from vimL functions.

bfredl commented:
I think it is RDY, if no one objects against `call_methods_atomic`

justinmk commented:
I would prefer call_atomic because, I think we should eliminate the arbitrary jargon distinction between "methods " and "functions".  It's confusing that we assigned special meaning there. Not all API functions are acting on an object in a meaningful sense, so it's more inside knowledge.

(In the tests I have been meaning to rename "meths" to "api" or something like that)

bfredl commented:
Well, the object in that case is the nvim instance, but I agree it is a bit internal jargon, so changing back to `call_atomic`.

justinmk commented on src/nvim/api/vim.c:
```diff
@@ -663,6 +663,113 @@ Array nvim_get_api_info(uint64_t channel_id)
+/// be indicated in the return value, see below.
+///
+/// @return an array with three elements. The first is the number of calls that
+/// where performed without errors. If it is the same as the length of "calls",
```
Wouldn't need the first item if we instead _always_ return the results.

justinmk commented on src/nvim/api/vim.c:
```diff
@@ -663,6 +663,113 @@ Array nvim_get_api_info(uint64_t channel_id)
+///
+/// @param calls an array of calls, where each call is described by an array
+/// with at least three elements: the request name, an array of arguments, and
+/// a boolean indicating if the return value should be sent back.
```
Could the third item (`retval` boolean) be a premature optimization? It would greatly simplify explanation/interface of `call_atomic` if we did not make results optional.

bfredl commented:
> (In the tests I have been meaning to rename "meths" to "api" or something like that)

:+1: I also think we should merge curbufmeths into curbuf so one could `curbuf.set_lines`

bfredl commented:
Ok, updated. It is probably simpler to make sure that that all methods where the return value is not essential, will not have a heavy return value, like we did for the `set_var` methods.

bfredl commented:
Did some more doc adjustments, if no more comments I will merge after tests.
